### PR TITLE
correct the method cache's usage of typemap simplesig

### PR DESCRIPTION
**Pull request by vtjnash**
fixes a performance issue found by @carnaval where we were continuously evicting the same entries from the cache. This likely shouldn't have been affecting too many places, but it was causing the misc test to fail the allocated test with no-inline, and may have slowed down `cconvert{T}(Type{T}, x::T)` also (compiling inference.ji seems to be a bit faster with this).

### CODE REVIEW

File: src/gf.c [hunk 1]
```diff
@@ -611,57 +609,73 @@ static jl_lambda_info_t *cache_method(jl_methtable_t *mt, union jl_typemap_t *ca
...
+
+    // compute the type this will be cached under
+    // if we haven't selected an origtype yet, promote `type`,
+    // and then decide if it is beneficial to build a new simplesig
+    if (origtype == type) {
+        type = NULL; // don't need `type` anymore: it's equivalent to the `origtype`
+        if (makesimplesig) {
+            // reduce the complexity of rejecting this entry in the cache
+            // by replacing non-simple types with jl_any_type to build a new `type`
+            // (the only case this applies to currently due to the above logic is jl_function_type)
```
Comment (tkelman) [inline src/gf.c]: in jl_function_type ?

File: src/gf.c [hunk 2]
```diff
@@ -611,57 +609,73 @@ static jl_lambda_info_t *cache_method(jl_methtable_t *mt, union jl_typemap_t *ca
...
+                jl_value_t *elt = jl_svecref(newparams, i);
+                if (elt == (jl_value_t*)jl_function_type)
+                    jl_svecset(newparams, i, jl_any_type);
+            }
+            type = jl_apply_tuple_type(newparams);
+            temp2 = (jl_value_t*)type;
+        }
+    }
+
+    jl_typemap_insert(cache, parent, origtype, jl_emptysvec, type, guardsigs, (jl_value_t*)newmeth, jl_cachearg_offset(mt), &lambda_cache, NULL);
```
Comment (tkelman) [inline src/gf.c]: line length

File: src/gf.c (hunk 1 above)
Comment (vtjnash) [inline src/gf.c]: garden path. the text above is correct, if a bit hard to follow.

Comment (tkelman) [inline src/gf.c]: please make it clearer

Comment (vtjnash) [inline src/gf.c]: the comment isn't new, just moved (I tried rewriting it too, but couldn't find a phrasing that was any better)
